### ISSUE TITLE
ExternalName service doesn't generate Environment Variables for service discovery

### ISSUE THREAD

**Issue by johngmyers** [kubernetes/kubernetes#60535]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): externalname service environment

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

(though it seems like a bug to me)

The externalName of ExternalName services should appear under the SERVICENAME_SERVICE_HOST environment variable in created containers.

**Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-09T21:51:06Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.5", GitCommit:"cce11c6a185279d037023e02ac5249e14daa22bf", GitTreeState:"clean", BuildDate:"2017-12-07T16:05:18Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**: (presuming irrelevant)

**What happened**:

root@debug-f95f68697-2cx2d:/# printenv
HOSTNAME=debug-f95f68697-2cx2d
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP_ADDR=10.0.0.1
KUBERNETES_PORT=tcp://10.0.0.1:443
PWD=/
HOME=/root
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP=tcp://10.0.0.1:443
TERM=xterm
SHLVL=1
KUBERNETES_SERVICE_PORT=443
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
KUBERNETES_SERVICE_HOST=10.0.0.1
_=/usr/bin/printenv

**What you expected to happen**:

Above should have additionally contained at least:

AUTHSERVICE_SERVICE_HOST=example.com

**How to reproduce it** (as minimally and precisely as possible):

kubectl create service externalname authservice --external-name example.com 
kubectl run -it debug --image debian:stretch

**Anything else do we need to know**:

Previously reported as #39356

**Comment by johngmyers** [comment]
@kubernetes/sig-network-feature-requests

**Comment by EdSchouten** [comment]
This issue is weakly related to a comment I just made in another bug: https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-375111483

Being able to use `ExternalName` to set environment variables to a string would be incredibly useful for pods that want to connect to the API server. Right now the API server tends to require a SSL certificate that has the IP address of the 'default/kubernetes' service as the subject name. If `ExternalName` would allow us to set `KUBERNETES_SERVICE_HOST` to the publicly used name of the API server, this requirement would be lifted entirely.

**Comment by miend** [comment]
+1, just ran into this while trying to use an ExternalName service to point to another service in a different namespace. This would be a very useful feature.

**Comment by wieslaw-gat** [comment]
+1 as well

**Comment by thockin** [comment]
In general, I'd reallllllly rather get rid of the env-var service exposure logic.  It's kind of a mess and a scalability problem.

ExternalName *does* produce a DNS record, which is more powerful than env var anyway.

If there's a good reason that doesn't work -- reopen?

**Comment by johngmyers** [comment]
The DNS record is not useful when it is a TLS service because the service will almost never have a matching SAN in its certificate. The code in the pod needs to dereference and use the underlying external domain

The documentation at https://kubernetes.io/docs/concepts/services-networking/service/ clearly states that the kubelet adds these environment variables "for each active Service". If ExternalName services are supposed to be excluded from this, then the documentation needs to be updated to specifically state this.

**Comment by asauber** [comment]
I have a need for this as well.

What if we made it optional for kubelet using an annotation similar to the following?

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    k8s.io/injectenv: always
```

**Comment by thockin** [comment]
Given that we want the env-vars to go away over time, I really really do not want to add this.  It's already a somewhat incompatible change (*_SERVICE_HOST has always been an IP, I can't say for sure whether it is safe or not to put a hostname).

What if we had a small sidecar container that watched a Service, extracted the name, and wrote it to a file, so you could share it via a volume?  Would that be an acceptable alternative?  I much prefer this model because it is FAR more general-purpose.

**Comment by johngmyers** [comment]
There is a variance between the documentation and the behavior of the code, which is by definition a bug. If the intention is to not bring the code into conformance with the documentation, then the documentation should be changed to accurately describe the intended behavior.

**Comment by johngmyers** [comment]
As I see it, the point of the env-vars interface is to get data into simple programs without having to put complicated logic inside the image. As the current implicit API appears to be problematic, perhaps an explicit API, for example adding a `serviceRef` field to `EnvFromSource`, would suit.

**Comment by asauber** [comment]
@thockin I would love to get rid of these environment variables, unfortunately, they are the [source of truth for client-go, and thus almost every Kubernetes controller in the wild](https://github.com/kubernetes/client-go/blob/9bbcc2938d41daa40d3080a1b6524afbe4e27bd9/rest/config.go#L456).

I work in an environment where for various reasons the Kube apiserver can only be reached by a DNS name. What I have to do for each cluster component is override `KUBERNETES_SERVICE_HOST` with a DNS names, which does work. My CNI driver, CSI driver, and even CoreDNS are configured this way. I would love it if `client-go` supported automatic discovery of the Kube apiserver using a DNS name, but it simply doesn't right now. The suggestion is how this can be implemented using the current APIs.

Obviously there is a bootstrapping problem before CoreDNS is healthy if we want to use `ExternalName` for the Kube apiserver. To overcome this I have `kubelet` and `kube-proxy` configured to reach the DNS name for the kube-apiserver (these are configured manually regardless, and this works perfectly). For `CoreDNS` I have overridden `KUBERNETES_SERVICE_HOST` with a DNS name, which also works perfectly.

The proposal for `KUBERNETES_SERVICE_URL` would help this situation somewhat, however **many controllers in the wild with "legacy" client-go** will expect to find the live host at `KUBERNETES_SERVICE_HOST`. This means that I currently have to instruct all of my users to override this environment variable with a DNS name for every helm chart or other application that acts as a Kubernetes controller. In general this is simply not practical.

Fixing this issue will solve my problems and many others who are in this situation.

In summary 1) It is safe to put at hostname in `KUBERNETES_SERVICE_URL`, 2) we should aim for a solution that will fix the problem of "legacy" `client-go` applications.

**Comment by thockin** [comment]
> As the current implicit API appears to be problematic, perhaps an explicit API, for example adding a serviceRef field to EnvFromSource, would suit.

This is not impossible, but frankly, every API feature we add comes with long-term cost to maintain, and they need to be worthwhile. Env vars are a terrible API for several reasons and I don't want to add API to make them even more complicated to use, especially for non-constant values.  I'd rather fix it as proposed in this issue or punt entirely to sidecars.

The benefit of a more general solution like a sidecar is that it solves for other problems, as well.  For concrete example, it has been asked how to project CRDs into env vars and files.  I am surely not going to endorse that as part of the core API.

> In summary 1) It is safe to put a DNS hostname in KUBERNETES_SERVICE_HOST

I missed the part where there was a solid argument establishing this as fact?  I have heard anecdotes and I suspect this is PROBABLY right, but I am not SURE of it.

@deads2k - is there any reason we can't fix client-go to fall back on DNS names?  That doesn't help the legacy clients, but it seems correct overall?

KUBERNETES_SERVICE_URL attempts to solve several problems, but ultimately has the same legacy client issue.

So, OK.  Let's assume for a minute that I cave on this.  Who is going to write the (small) KEP?

**Comment by asauber** [comment]
@thockin I'd be happy to write up the KEP

**Comment by aojea** [comment]
Have you tried [Pod Presets](https://kubernetes.io/docs/concepts/workloads/pods/podpreset/) for injecting environment variables? 

I think that this can solve some of the use cases explained in this issue
 https://kubernetes.io/docs/tasks/inject-data-application/podpreset/#use-pod-presets-to-inject-environment-variables-and-volumes

**Comment by asauber** [comment]
@aojea I haven't tried Pod Presets for my use case. I'll give them a spin and see how it goes.

**Comment by asauber** [comment]
I tested out Pod Presets, and found two issues with them for my use case.

1. They're namespaced, which doesn't give me cluster-wide env vars, which is a requirement in my case.
2. They don't appear to work with init containers, even though that was supposedly fixed a while back.

The next thing that I will try is a [Mutating Admission Webhook](https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/), and see if I can get the functionality that I need. If this also doesn't appear viable, then I am still interested in writing up the KEP for this issue.

**Comment by asauber** [comment]
Happy to report that injecting the relevant environment variables via a Mutating Admission Webhook does the trick. All controllers and operators that I use have been able to use `KUBERNETES_SERVICE_HOST` as a hostname.

However this seems like quite a hack, and I would prefer that ExternalName services have this behavior.

**Comment by nurhun** [comment]
Any updates if this will be considered
